### PR TITLE
adds task and patient projections

### LINKED ISSUE #685: Postgres projection
### Describe the feature

Implement an example read model via a custom projection for the `tasks-svc`. Utilize our existing SQLc architecture to implement the models like `task`, `subtask`, `patient` ... . The state of these models progresses through our events via projections.

The Postgres projection should mostly live in `/services/tasks-svc/internal/postgres-projection/`, please check out if we need to have the SQLc setup in the root directory of the service (due to our tooling) or if can easily move them into subdirectories as presented.

### Why is this feature needed?

We discussed an approach to implement our read models with our already developed `hwes.CustomProjection` and is basically a merge of our Postgres (SQLc) and Event Sourcing approach.

### Additional context

In the future, we will probably have more projections for different ways of displaying our data.

### PR DESCRIPTION
**Pull request by PaulKalho**
## Which issues does this pull request close?
closes #685

### CODE REVIEW

**Comment by MaxSchaefer** [comment]
> @MaxSchaefer @FoseFx and again no idea why pipeline fails with `ERROR: permission denied for table schema_migrations`

Let me check this. I guess we need to give the db ci user access to the two new databases.

**Comment by MaxSchaefer** [comment]
@PaulKalho permissions problems solved. Now the CI reports the correct problem. You cannot modify existing migrations.

File: services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	p.RegisterEventListener(patientEventsV1.PatientDischarged, p.onPatientDischarged)
+	p.RegisterEventListener(patientEventsV1.NotesUpdated, p.onNotesUpdated)
+	p.RegisterEventListener(patientEventsV1.HumanReadableIdentifierUpdated, p.onHumanReadableIdentifierUpdated)
+	p.RegisterEventListener(patientEventsV1.PatientReadmitted, p.onPatientReadmitted)
+}
+
+// Event handlers
+func (a *Projection) onPatientCreated(ctx context.Context, evt hwes.Event) (error, esdb.Nack_Action) {
+	log := zlog.Ctx(ctx)
+	patientRepo := patient_repo.New(hwdb.GetDB())
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go:42]: might make sense to move the repos into the Projection (`a.patientRepo`)

File: services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+	var payload patientEventsV1.PatientCreatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		log.Error().Err(err).Msg("unmarshal failed")
+		return err, esdb.Nack_Retry
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go:47]: when the payload is invalid, should we retry? I'm not sure about this either way

File: services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
+package patient_postgres_projection
+
+import (
+	"context"
+	"fmt"
+	"github.com/EventStore/EventStore-Client-Go/esdb"
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go:6]: wrong esdb, use v4 now

File: services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+func (a *Projection) onBedUnassigned(ctx context.Context, evt hwes.Event) (error, esdb.Nack_Action) {
+	patientRepo := patient_repo.New(hwdb.GetDB())
+
+	timestamp := pgtype.Timestamp{Time: evt.Timestamp, Valid: true}
+	if err := patientRepo.UpdatePatientBedId(ctx, patient_repo.UpdatePatientBedIdParams{
+		ID:        evt.AggregateID,
+		BedID:     uuid.NullUUID{},
+		UpdatedAt: timestamp,
+	}); err != nil {
+		return nil, esdb.Nack_Retry
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go:105]: log errors like these

File: services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+func (a *Projection) onPatientReadmitted(ctx context.Context, evt hwes.Event) (error, esdb.Nack_Action) {
+	patientRepo := patient_repo.New(hwdb.GetDB())
+
+	timestamp := pgtype.Timestamp{Time: evt.Timestamp, Valid: true}
+	isDischarged := int32(0)
+	if err := patientRepo.UpdatePatient(ctx, patient_repo.UpdatePatientParams{
+		ID:           evt.AggregateID,
+		IsDischarged: &isDischarged,
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go:178]: [hwutil.PtrTo](https://github.com/helpwave/services/blob/main/libs/hwutil/parse.go#L101C6-L101C11)

File: services/tasks-svc/repos/patient_repo.sql [hunk 6]
```diff
@@ -0,0 +1,18 @@
+-- name: CreatePatient :exec
+INSERT INTO patients
+	(id, human_readable_identifier, notes, created_at, updated_at)
+VALUES ($1, $2, $3, $4, $5);
+
+-- name: UpdatePatient :exec
+UPDATE patients
+SET human_readable_identifier = coalesce(sqlc.narg('human_readable_identfier'), human_readable_identifier),
+    notes = coalesce(sqlc.narg('notes'), notes),
+	updated_at = coalesce(sqlc.narg('updated_at'), updated_at),
```
Comment (FoseFx) [inline services/tasks-svc/repos/patient_repo.sql:10]: formatting

File: services/tasks-svc/migrations/000020_organization_ids.up.sql [hunk 7]
```diff
@@ -0,0 +1,17 @@
+ALTER TABLE patients
+	DROP COLUMN organization_id;
```
Comment (FoseFx) [inline services/tasks-svc/migrations/000020_organization_ids.up.sql:2]: IF EXISTS

migrations should be idempotent

File: services/tasks-svc/migrations/000020_organization_ids.down.sql [hunk 8]
```diff
@@ -0,0 +1,22 @@
+ALTER TABLE patients
+	ADD COLUMN IF NOT EXISTS
+	organization_id uuid NOT NULL;
```
Comment (FoseFx) [inline services/tasks-svc/migrations/000020_organization_ids.down.sql:3]: I doubt you can just add a column and constrain it NOT NULL without providing a value to set

File: services/tasks-svc/repos/patient_repo.sql (hunk 6 above)
Comment (PaulKalho) [reply services/tasks-svc/repos/patient_repo.sql:10]: Seems to be a GitHub displaying error, is not indented in the file itself

File: services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go (hunk 2 above)
Comment (PaulKalho) [reply services/tasks-svc/internal/patient/projections/patient_postgres_projection/patient_postgres_projection.go]: I was not quite sure about this one either. I've talked with @MaxSchaefer about the `NackActions` and we decided retries are sufficient for our current state of `tasks-svc`

File: services/tasks-svc/repos/patient_repo.sql (hunk 6 above)
Comment (FoseFx) [reply services/tasks-svc/repos/patient_repo.sql:10]: check the whitespace of the notes line

**Review by FoseFx — APPROVED** [review]

**Pull request merged**